// Copyright 2025 TIER IV, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef AUTOWARE__CUDA_POINTCLOUD_PREPROCESSOR__COMMON_KERNELS_HPP_
#define AUTOWARE__CUDA_POINTCLOUD_PREPROCESSOR__COMMON_KERNELS_HPP_

#include "autoware/cuda_pointcloud_preprocessor/point_types.hpp"
#include "autoware/cuda_pointcloud_preprocessor/types.hpp"

#include <cuda_runtime.h>

namespace autoware::cuda_pointcloud_preprocessor
{
void transformPointsLaunch(
  const InputPointType * input_points, InputPointType * output_points, int num_points,
  TransformStruct transform, int threads_per_block, int blocks_per_grid, cudaStream_t & stream);

void cropBoxLaunch(
  InputPointType * d_points, std::uint32_t * output_crop_mask, std::uint8_t * output_nan_mask,
  int num_points, const CropBoxParameters * crop_box_parameters_ptr, int num_crop_boxes,
  int threads_per_block, int blocks_per_grid, cudaStream_t & stream);

void combineMasksLaunch(
  const std::uint32_t * mask1, const std::uint32_t * mask2, int num_points,
  std::uint32_t * output_mask, int threads_per_block, int blocks_per_grid, cudaStream_t & stream);

void extractPointsLaunch(
  InputPointType * input_points, std::uint32_t * masks, std::uint32_t * indices, int num_points,
  OutputPointType * output_points, int threads_per_block, int blocks_per_grid,
  cudaStream_t & stream);

void extractInputPointsLaunch(
  InputPointType * input_points, std::uint32_t * masks, std::uint32_t * indices, int num_points,
  InputPointType * output_points, int threads_per_block, int blocks_per_grid,
  cudaStream_t & stream);

void convertPointXYZIRCToInputPointTypeLaunch(
  const OutputPointType * input_points, InputPointType * output_points, int num_points,
  int threads_per_block, int blocks_per_grid, cudaStream_t & stream);

// Pointcloud to costmap kernels
struct CostmapGridParams
{
  float grid_length_x;
  float grid_length_y;
  float grid_resolution;
  float grid_position_x;
  float grid_position_y;
  float maximum_height_thres;
  float minimum_height_thres;
  float grid_min_value;
  float grid_max_value;
  int32_t grid_size_x;
  int32_t grid_size_y;
};

void assignPointsToGridCellsLaunch(
  const InputPointType * input_points, int32_t * grid_indices, int num_points,
  const CostmapGridParams * params, int threads_per_block, int blocks_per_grid,
  cudaStream_t & stream);

void calculateCostmapFromGridCellsLaunch(
  const InputPointType * input_points, const int32_t * grid_indices, float * costmap_data,
  int num_points, const CostmapGridParams * params, int threads_per_block, int blocks_per_grid,
  cudaStream_t & stream);

}  // namespace autoware::cuda_pointcloud_preprocessor

#endif  // AUTOWARE__CUDA_POINTCLOUD_PREPROCESSOR__COMMON_KERNELS_HPP_
